---
title: "Transmission analysis"
author: "Karim"
output:
  html_document: default
  
---

### **1. data quality control**     
This includes SNPs filtration based on their MAF (minor allele frequencies), their percentage of missing genotypes (missingness) as well as the filtration of isolates based on missingness.    
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Filter_data.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the input VCF file_     
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the MAF cut-off_     
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the missingness cut-off_     
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the path to the folder where to store the output files_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the numder of threads_    
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _2 `pdf` files (plots of SNPs and isolates missingness and MAF) and a text file named `MyGenotype.txt`_       
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**        
```{bash eval = FALSE}
Rscript Filter_data.R Data/file.vcf.gz 0.01 20 Results/ 15      
```

***         

### **2. estimation of the complexity of infection (COI)**
The input VCF file will be filtered so that it contains only the SNPs and isolates that have passed the QC step. The filtered VCF file is then used to compute Fws (within sample fixation index) using the `moimix` R package.   
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Compute_COI.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the filtered genotype file_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the input VCF file_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the sample metadata file_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the file that contains the order in which you wish your population to be shown in the output plot, 1 population per line_        
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _the filtered VCF file (file with a prefix `Passed_QC_genotypes`), 2 `pdf` files (Fws plots)_       
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**        
```{bash eval = FALSE}
Rscript Compute_COI.R Results/MyGenotype.txt Data/MyDelgeme4.vcf.gz Data/SampleMetadata.txt Data/Pop_Order.txt    
```

***  

### **3. construction of the input files for relatedness inference ** 
11 regions of the genome that are associated with drug resistance are first discarded. Then the reference alleles are recoded as `0`, the alternate alleles as `1`. We defined 3 methods of recoding the mixed genotypes:   
&nbsp;&nbsp;&nbsp;&nbsp; mixed genotypes replaced by the minor allele calculated across all samples for a given SNPs (`minor_David`)   
&nbsp;&nbsp;&nbsp;&nbsp; mixed genotypes considered as a third allele and recoded as 2 (`raw`)    
&nbsp;&nbsp;&nbsp;&nbsp; mixed genotypes replaced by the minor allele calculated based on the allelic depth on a given isolates at a particular SNP (`minor_Karim`)   
Note that for the first and third method, when the frequencies of the 2 alleles are the same, it is recoded based on a Bernoulli distribution.   
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Construct_Genotype_File.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the filtered VCF file_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the output files directory_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the file with sample metadata file_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _full path to the file with genomic locations of the regions to be discarded_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the mixed genotypes recoding method _    
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _creates `RDS` files with the recoded genotype data, 1 for each population_    
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**           
```{bash eval = FALSE}
Rscript Construct_Genotype_File.R Results/Passed_QC_genotypes.vcf.gz Results/GenotypeFiles Data/SampleMetadata.txt Data/Resistance_Gene.txt raw   
```

***  

### **4. Inference of relatedness between all pairs of populations ** 
We estimated relatedness between pairs of isolates from 2 distinct populations based on the method proposed by Aimee Taylor and Coauthors https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1009101   
Between every pairs of population, the relatedness is evaluated using data from 5 MAF cut-offs (set of SNPs with MAF>0%, MAF>10%, MAF>20%, MAF>30%, MAF>40%)   
Before running this stage, make sure the `runSim.sh` is edited according to your HPC specifications.   
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Run_Relatedness.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the folder with the recoded data_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the output files directory_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the file with population order_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the mixed genotypes recoding method_     
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _creates `RDS` files within different folders, 1 for each pair of populations_    
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**        
```{bash eval = FALSE}
Rscript Run_Relatedness.R /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/Raw_Genotypes /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Data/Pop_Order.txt raw      
```

***  

### **5. Summarizing the relatedness files from each pairs of populations ** 
When estimating relatedness, pairs of samples from 2 populations are splitted and stored into separated RDS files. Here, we will concatenate these files and form one RDS file for each MAF cut-off.  
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Summarize_IBD_Files.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the input folder_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the output files directory_    
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _creates `RDS` files within output folders_    
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**        
```{bash eval = FALSE}
Rscript Summarize_IBD_Files.R /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder/From_raw_Data /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder/From_raw_Data/Summarized  
```

***

### **6. Determination of the percentage of highly related infections pairs **   
We set a relatedness cut-off (in our case, we varied the cut-off from 0.2 to 0.6 by step of 0.1) and considered any pairs of infection with a relatedness value greater than or equal to this cut-off as a highly related pair of infections. If your samples were collected in different years, the relatedness data will be summarized in a way that for pair of populations, relatedness values will be reported between the different sampling years.  
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Identify_HRPI.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the folder with the summarized data_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the file with the sample metadata file_    
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _creates several `RDS` files within output folders. Among these, will be interested in the once with a suffix `_All_Pairs.RDS`_        
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**          
```{bash eval = FALSE}
Rscript Identify_HRPI.R /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder/From_raw_Data/Summarized /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Data/SampleMetadata.txt  
```

***

### **7. Combine relatedness files **    
Files from the above step (6) will be combined into one single file that will be used for the subsequent analysis. Note that the combined output file is a list of 5 data frames that contain the same information. You can chose any of them for the next stages.    
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Combine_Relatedness_Files.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the folder with the summarized data_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the file with the physical distances between the sampling sites_    
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _creates `RDS` files within a folder named `Combined/`_    
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**        
```{bash eval = FALSE}
Rscript Combine_Relatedness_Files.R /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder/From_raw_Data/Summarized /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Data/PhysicalDistanceBetweenLocations.txt    
```


***

### **8. Analysis section **    
###### **8.1. get the file to use for analysis       
```{r echo = FALSE, eval = TRUE}
#for raw data 
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
system(sprintf("mkdir -p %s", relatednessDir)) 
relatedness = readRDS(paste0(relatednessDir,"/Combined_0.6.RDS"))   
allPairs_From_Raw = relatedness[[1]]   
saveRDS(allPairs_From_Raw, paste0(relatednessDir,"/All_Relatedness_From_Raw.RDS"))    

#for minor_David data  
#relatedness = readRDS("/lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder/From_minor_#David_Data/Summarized/Combined/Combined_0.6.RDS")   
#allPairs_From_Raw = relatedness[[1]]   
#saveRDS(allPairs_From_Raw, paste0(relatednessDir,"/All_Relatedness_From_Minor_David.RDS")) 

#for minor_Karim data   
#relatedness = readRDS("/lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/IBD_folder/From_minor_#Karim_Data/Summarized/Combined/Combined_0.6.RDS")   
#allPairs_From_Raw = relatedness[[1]]   
#saveRDS(allPairs_From_Raw, paste0(relatednessDir,"/All_Relatedness_From_Minor_Karim.RDS")) 
```

###### **8.2. determining MAF cut-off (by comparing the results from the different mixed genotypes recoding methods against the results from raw data)       
```{r echo = FALSE, eval = FALSE}
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
raw = readRDS(paste0(relatednessDir,"/All_Relatedness_From_Raw.RDS"))    
minorDavid = readRDS(paste0(relatednessDir,"/All_Relatedness_From_Minor_David.RDS"))    
minorKarim = readRDS(paste0(relatednessDir,"/All_Relatedness_From_Minor_Karim.RDS"))     

#between raw and minorDavid
raw = as.data.table(raw); setkeyv(raw,c('iid1','iid2'))  
minorDavid = as.data.table(minorDavid); setkeyv(minorDavid,c('iid1','iid2'))     
joined = raw[minorDavid, nomatch=0]  
joined$`rMaf>0` = as.numeric(joined$`rMaf>0`)       
joined$`i.rMaf>0` = as.numeric(joined$`i.rMaf>0`)      
p1 = ggplot(joined, aes(x=joined$`rMaf>0`, y=joined$`i.rMaf>0`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>0%; correlation coefficient=",round(cor(joined$`rMaf>0`, joined$`i.rMaf>0`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>10` = as.numeric(joined$`rMaf>10`)       
joined$`i.rMaf>10` = as.numeric(joined$`i.rMaf>10`)      
p2 = ggplot(joined, aes(x=joined$`rMaf>10`, y=joined$`i.rMaf>10`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>10%; correlation coefficient=",round(cor(joined$`rMaf>10`, joined$`i.rMaf>10`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>20` = as.numeric(joined$`rMaf>20`)       
joined$`i.rMaf>20` = as.numeric(joined$`i.rMaf>20`)      
p3 = ggplot(joined, aes(x=joined$`rMaf>20`, y=joined$`i.rMaf>20`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>20%; correlation coefficient=",round(cor(joined$`rMaf>20`, joined$`i.rMaf>20`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>30` = as.numeric(joined$`rMaf>30`)       
joined$`i.rMaf>30` = as.numeric(joined$`i.rMaf>30`)      
p4 = ggplot(joined, aes(x=joined$`rMaf>30`, y=joined$`i.rMaf>30`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>30%; correlation coefficient=",round(cor(joined$`rMaf>30`, joined$`i.rMaf>30`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>40` = as.numeric(joined$`rMaf>40`)       
joined$`i.rMaf>40` = as.numeric(joined$`i.rMaf>40`)      
p5 = ggplot(joined, aes(x=joined$`rMaf>40`, y=joined$`i.rMaf>40`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>40%; correlation coefficient=",round(cor(joined$`rMaf>40`, joined$`i.rMaf>40`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))

pdf(paste0(relatednessDir,"/raw_vs_minorDavid.pdf"))
par(mfrow=c(3,2))
print(p1); print(p2); print(p3); print(p4); print(p5)
dev.off()


#between raw and minorKarim
raw = as.data.table(raw); setkeyv(raw,c('iid1','iid2'))  
minorKarim = as.data.table(minorKarim); setkeyv(minorKarim,c('iid1','iid2'))     
joined = raw[minorKarim, nomatch=0]  
joined$`rMaf>0` = as.numeric(joined$`rMaf>0`)       
joined$`i.rMaf>0` = as.numeric(joined$`i.rMaf>0`)      
p1 = ggplot(joined, aes(x=joined$`rMaf>0`, y=joined$`i.rMaf>0`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>0%; correlation coefficient=",round(cor(joined$`rMaf>0`, joined$`i.rMaf>0`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>10` = as.numeric(joined$`rMaf>10`)       
joined$`i.rMaf>10` = as.numeric(joined$`i.rMaf>10`)      
p2 = ggplot(joined, aes(x=joined$`rMaf>10`, y=joined$`i.rMaf>10`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>10%; correlation coefficient=",round(cor(joined$`rMaf>10`, joined$`i.rMaf>10`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>20` = as.numeric(joined$`rMaf>20`)       
joined$`i.rMaf>20` = as.numeric(joined$`i.rMaf>20`)      
p3 = ggplot(joined, aes(x=joined$`rMaf>20`, y=joined$`i.rMaf>20`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>20%; correlation coefficient=",round(cor(joined$`rMaf>20`, joined$`i.rMaf>20`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>30` = as.numeric(joined$`rMaf>30`)       
joined$`i.rMaf>30` = as.numeric(joined$`i.rMaf>30`)      
p4 = ggplot(joined, aes(x=joined$`rMaf>30`, y=joined$`i.rMaf>30`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>30%; correlation coefficient=",round(cor(joined$`rMaf>30`, joined$`i.rMaf>30`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))
    
joined$`rMaf>40` = as.numeric(joined$`rMaf>40`)       
joined$`i.rMaf>40` = as.numeric(joined$`i.rMaf>40`)      
p5 = ggplot(joined, aes(x=joined$`rMaf>40`, y=joined$`i.rMaf>40`)) + 
    geom_point(shape=19, size=0.1) +
    geom_smooth(method=lm,  linetype="dashed",color="darkred", fill="blue") +
    ggtitle(paste0("raw vs minor_David; MAF>40%; correlation coefficient=",round(cor(joined$`rMaf>40`, joined$`i.rMaf>40`), digits = 2))) +
    xlab("raw") +ylab("minor_David") +
    theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()
    ) +
    theme(axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))

pdf(paste0(relatednessDir,"/raw_vs_minorKarim.pdf"))
par(mfrow=c(3,2))
print(p1); print(p2); print(p3); print(p4); print(p5)
dev.off()

```

The results from the above showed that the raw data correlate more with the data from minor_David. It also shows that the highest correlations between the raw and minor_David are from MAF>0 and MAF>30. For the subsequent analysis, we choose to use the relatedness calculated from the raw data with MAF>30. We also recommend that when need is to use the data where the mixed genotypes are recoded based on minor_David.   

***

### **9. Convert relatedness data frame into matrix **    
In order to plot the relatedness between 1 isolate and others, the data needs to be converted into a matrix.  
&nbsp;&nbsp;&nbsp;&nbsp;**Function name:** _Create_relatedness_matrix.R_    
&nbsp;&nbsp;&nbsp;&nbsp;**Parameters:**    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the relatedness file_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the full path to the file sample metadata_  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the MAF cut-off_   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the path to the output folder_    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- _the output file prefix_  
&nbsp;&nbsp;&nbsp;&nbsp;**Outputs:** _creates an `RDS` file with the prefix chosen by the user_     
&nbsp;&nbsp;&nbsp;&nbsp;**Example:**          
```{bash eval = FALSE}
Rscript Create_relatedness_matrix.R /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/RelatednessFolder/All_Relatedness_From_Raw.RDS /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Data/SampleMetadata.txt 0.3 /lustre/scratch117/casm/team273/km28/Karim/transmission_analysis/Results/RelatednessFolder rMatrixFrom_Raw
```

###### **9.1. drawing the mean relatedness as a heatmap **   
```{r echo = FALSE} 
library(ggplot2)
library(reshape2)
library(data.table)
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
rMatrix = readRDS(paste0(relatednessDir,'/rMatrixFrom_Raw.RDS'))
meta=fread(paste0(relatednessDir,"/SampleMetadata.txt"))
loc = unique(meta$Location)
matIBDseg = matrix(0, nrow = length(loc), ncol = length(loc))
newOrder = c('Serrekunda','Brikama','Besse','Chogen','Farafenni','DongoroBa','Basse','SareWuro','Njayel','Bounkiling','Thies')
isolateNumber = c(36,230,17,8,157,11,117,62,80,59,92)

meta$SampleID = as.character(meta$SampleID)
rownames(matIBDseg) = newOrder
colnames(matIBDseg) = newOrder #rownames(matIBDseg) #= loc
for(i in 1:length(newOrder))
{
    t1 = newOrder[i]
    mm = meta[which(meta$Location==t1),]$SampleID
    replaceSep=function(x){gsub('-','.',x)}
    mm = unlist(lapply(mm, replaceSep))
    m1 = match(mm, rownames(rMatrix))
    m3 = match(mm, colnames(rMatrix))
    #m1 = which(meta$Location==t1)
    for(j in 1:length(loc))
    {
        t2=newOrder[j]
        mm = meta[which(meta$Location==t2),]$SampleID
        m2 = match(mm, colnames(rMatrix))
        m4 = match(mm, rownames(rMatrix))
        m5 = c(m1,m4); m6=c(m2,m3)
        target = rMatrix[m5, m6]
        matIBDseg[i,j] = mean(as.vector(target), na.rm = TRUE)
        
    }
}

meltedMatIBD = reshape2::melt(matIBDseg, na.rm = TRUE)
meltedMatIBD$Var1=as.factor(meltedMatIBD$Var1)
meltedMatIBD$Var2=as.factor(meltedMatIBD$Var2)
#meltedMatIBD$isolates = NA
#for(i in 1:length(newOrder))
#{
#    print(i)
#    meltedMatIBD[which(meltedMatIBD$Var1==newOrder[i] & meltedMatIBD$Var2==newOrder[i]),]$isolates = isolateNumber[i]
#}
#ggplot(data = meltedMatIBD, aes(Var2, Var1, fill = value))+
#    geom_tile(color = "white")+
#    scale_fill_gradient2(low = "red", high = "green", mid = "yellow",
#                         midpoint = 0.1, limit = c(min(meltedMatIBD$value),0.15), space = "Lab",name="mean IBD") +
#    theme_minimal()+
#    theme(axis.text.x = element_text(angle = 90, vjust = 1, size = 8, hjust = 1))+
#    coord_fixed()+
#    labs(title = "mean IBD") +
#    theme(
#        axis.title.x = element_blank(),
#        axis.title.y = element_blank(),
#        panel.grid.major = element_blank(),
#        panel.border = element_blank(),
#        panel.background = element_blank(),
#        axis.ticks = element_blank()) #+
    # geom_text(aes(Var2, Var1, label = isolates), color = "black", size = 4)


get_upper_tri = function(cormat)
{
    cormat[upper.tri(cormat)] = NA
    return(cormat)
}
upper_tri <- get_upper_tri(matIBDseg)
meltedMatIBD = reshape2::melt(upper_tri, na.rm = TRUE)
meltedMatIBD$isolates = NA
#for(i in 1:length(newOrder))
#{
#    if(i==11)
#        next
#    meltedMatIBD[which(meltedMatIBD$Var1==newOrder[i] & #meltedMatIBD$Var2==newOrder[i]),]$isolates = isolateNumber[i]
#}

meltedMatIBD$Var1=as.factor(meltedMatIBD$Var1)
meltedMatIBD$Var2=as.factor(meltedMatIBD$Var2)
ggplot(data = meltedMatIBD, aes(Var2, Var1, fill = value))+
    geom_tile(color = "white")+
    scale_fill_gradient2(low = "red", high = "green", mid = "yellow",
                         midpoint = 0.1, limit = c(min(meltedMatIBD$value),max(meltedMatIBD$value)), space = "Lab",name="mean") +
    theme_minimal()+
    theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 8, hjust = 1))+
    coord_fixed()+
    labs(title = "") +
    theme(
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.ticks = element_blank(),
        legend.justification = c(1, 0),
        legend.position = c(1.1, 0.2),
        legend.direction = "vertical") #+
    #geom_text(aes(Var2, Var1, label = isolates), color = "black", size = 4) 


```

##### **9.2. population structure**
```{r echo=FALSE}
library(umap)
library(data.table)
library(ggplot2)
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
rMatrix = readRDS(paste0(relatednessDir,'/rMatrixFrom_Raw.RDS'))
meta=fread(paste0(relatednessDir,"/SampleMetadata.txt"))
modifier = function(x){gsub('-','.',x)}
meta$SampleID=lapply(meta$SampleID,modifier)

rMatrix[is.na(rMatrix)]=0
kk = umap(rMatrix)
kk = kk$layout
m=match(rownames(kk), meta$SampleID)
kk = cbind(kk, meta$Location[m])
kk = as.data.frame(kk, stringsAsFactors = FALSE)
kk$V1=as.numeric(kk$V1)
kk$V2=as.numeric(kk$V2)
kk$V3=as.factor(kk$V3)
rownames(kk)=NULL
ggplot(kk, aes(x=V1, y=V2, color=V3)) + 
  geom_point() + 
  scale_color_manual(name="Location",values=c("steelblue", "red",     "magenta","blue","black","green","yellow","gray","cyan","purple","orange")) +
  theme_bw()+
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = 60, vjust = 0.5))+
  theme( axis.line = element_line(colour = "gray", size = 0.5, linetype = "solid")) +
  xlab("PC1") + ylab("PC2")
```




###### **9.3. determination of the proportion of HRI between pairs of populations **   
```{r echo = FALSE}
library(data.table)
require(knitr)
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
raw = readRDS(paste0(relatednessDir,"/All_Relatedness_From_Raw.RDS"))
raw$`rMaf>30` = as.numeric(raw$`rMaf>30`)
hri = raw[which(raw$`rMaf>30`>=0.6),]

setkeyv(hri, c("site1","site2"))
nhri = hri[,.N,by=c("site1","site2")]
nhri$region1 = ""; nhri$region2=""
nhri[which(nhri$site1=="Basse" | nhri$site1=="SareWuro" | nhri$site1=="Njayel"),]$region1="east"
nhri[which(nhri$site1=="Farafenni" | nhri$site1=="DongoroBa" | nhri$site1=="Chogen"),]$region1="center"
nhri[which(nhri$site1=="Besse" | nhri$site1=="Brikama"),]$region1="west"
nhri[which(nhri$site1=="Bounkiling" | nhri$site1=="Thies"),]$region1="sen"

nhri[which(nhri$site2=="Basse" | nhri$site2=="SareWuro" | nhri$site2=="Njayel"),]$region2="east"
nhri[which(nhri$site2=="Farafenni" | nhri$site2=="DongoroBa" | nhri$site2=="Chogen"),]$region2="center"
nhri[which(nhri$site2=="Besse" | nhri$site2=="Brikama"),]$region2="west"
nhri[which(nhri$site2=="Bounkiling" | nhri$site2=="Thies"),]$region2="sen"
setkeyv(nhri, c("region1","region1"))
phri = nhri[,sum(N),by=c("region1","region2")]
phri$pAgainstHRI=round(phri$V1/sum(phri$V1), 3)


setkeyv(raw, c("site1","site2"))
rawByLoc = raw[,.N,by=c("site1","site2")]
#getting the proportion of HRI pairs between location
setkeyv(rawByLoc, c("site1","site2"))
setkeyv(nhri, c("site1","site2"))
nhri = nhri[rawByLoc, nomatch=0]
names(nhri)[3]="#HRI"; names(nhri)[6]="#totalPairs"; nhri=subset(nhri, select = -c(7,8))
nhri$pHRI = nhri$`#HRI`/nhri$`#totalPairs`
kable(nhri, caption="proportion of HRI between locations")
saveRDS(nhri,paste0(relatednessDir,'/Loc_PropHRI.RDS'))

#getting the proportion of HRI between regions
rawByLoc$region1 = ""; rawByLoc$region2=""
rawByLoc[which(rawByLoc$site1=="Basse" | rawByLoc$site1=="SareWuro" | rawByLoc$site1=="Njayel"),]$region1="east"
rawByLoc[which(rawByLoc$site1=="Farafenni" | rawByLoc$site1=="DongoroBa" | rawByLoc$site1=="Chogen"),]$region1="center"
rawByLoc[which(rawByLoc$site1=="Besse" | rawByLoc$site1=="Brikama" | rawByLoc$site1=="Serrekunda"),]$region1="west"
rawByLoc[which(rawByLoc$site1=="Bounkiling" | rawByLoc$site1=="Thies"),]$region1="sen"

rawByLoc[which(rawByLoc$site2=="Basse" | rawByLoc$site2=="SareWuro" | rawByLoc$site2=="Njayel"),]$region2="east"
rawByLoc[which(rawByLoc$site2=="Farafenni" | rawByLoc$site2=="DongoroBa" | rawByLoc$site2=="Chogen"),]$region2="center"
rawByLoc[which(rawByLoc$site2=="Besse" | rawByLoc$site2=="Brikama" | rawByLoc$site2=="Serrekunda"),]$region2="west"
rawByLoc[which(rawByLoc$site2=="Bounkiling" | rawByLoc$site2=="Thies"),]$region2="sen"
setkeyv(rawByLoc, c("region1","region1"))
praw = rawByLoc[,sum(N),by=c("region1","region2")]
phri$pAgainstAll=round(phri$V1/c(25890,39109,42733,51316,84381,13912,3136,47726,5428), 5)
saveRDS(phri,paste0(relatednessDir,'/Region_PropHRI.RDS'))

names(phri)[3] = 'N'
kable(phri, caption="proportion of HRI between regions")
```




### **10. spatial analysis (drawing connectivity using the proportion of HRI) **     
```{r echo=FALSE}
suppressPackageStartupMessages(expr = "")
require("raster")
require("rgdal")
require("dplyr")
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
relatednessTable = readRDS(paste0(relatednessDir,'/Loc_PropHRI.RDS'))
relatednessTable$LineWidth = relatednessTable$pHRI*150
gps = fread(paste0(relatednessDir,"/GPScoordinates.txt"))
gmb = getData(name = "GADM", country = "GMB", level=0)


withiGambia1 = relatednessTable[which(relatednessTable$site1!='Thies' & relatednessTable$site1!='Bounkiling'), ]
withiGambia2 = withiGambia1[which(withiGambia1$site2!='Thies' & withiGambia1$site2!='Bounkiling'), ]
#withiGambia = distinct(rbind(withiGambia1, withiGambia2))
gambia = gps[which(gps$Village!='Thies' & gps$Village!='Bounkiling' & gps$Village!='Velingara' & gps$Village!='SareSeedy' & gps$Village!='Serrekunda'),]

plot(gmb)
points(x=gambia$longitude, y=gambia$latitude, pch=19, cex=1.5, col=c('black','black','blue','blue','green','green','green','blue')) 

#for chogen
chogenAllFiles = list.files(path = paste0(relatednessDir,"/SiteConnectors"), pattern = "Chogen_", full.names = TRUE)
for(i in 1:length(chogenAllFiles))
{
    if(i==6 | i==8) #| i==8
    {
        s = gsub('.csv','',basename(chogenAllFiles[i]))
        s = unlist(strsplit(s,'_'))
        t = fread(chogenAllFiles[i])
        r = withiGambia2[which(withiGambia2$site1==s[1] & withiGambia2$site2==s[2]),]$LineWidth
        if(length(r) == 0)
            r = withiGambia2[which(withiGambia2$site1==s[2] & withiGambia2$site2==s[1]),]$LineWidth
        if(i==6)
          lines(x=t$long[2000:4002], y=t$lat[2000:4002], col='steelblue', lwd=r)
        else
            lines(x=t$long[1:2000], y=t$lat[1:2000], col='steelblue', lwd=r)
    }
}

#for Basse
basseAllFiles = list.files(path = paste0(relatednessDir,"/SiteConnectors"), pattern = "Basse_", full.names = TRUE)
for(i in 1:length(basseAllFiles))
{
    if( i==8 | i==10 ) #| i==2 | i==5 | i==11 | i==7
    {
        s = gsub('.csv','',basename(basseAllFiles[i]))
        s = unlist(strsplit(s,'_'))
        t = fread(basseAllFiles[i])
        r = withiGambia2[which(withiGambia2$site1==s[1] & withiGambia2$site2==s[2]),]$LineWidth
        if(length(r) == 0)
            r = withiGambia2[which(withiGambia2$site1==s[2] & withiGambia2$site2==s[1]),]$LineWidth
        lines(x=t$long[1:2000], y=t$lat[1:2000], col='steelblue', lwd=r)
    }
}

#for Besse
besseAllFiles = list.files(path = paste0(relatednessDir,"/SiteConnectors"), pattern = "Besse_", full.names = TRUE)
for(i in 1:length(besseAllFiles))
{
    if(i==5 | i==6 | i==8 | i==10) #| i==8 | i==4
    {
        s = gsub('.csv','',basename(besseAllFiles[i]))
        s = unlist(strsplit(s,'_'))
        t = fread(besseAllFiles[i])
        r = withiGambia2[which(withiGambia2$site1==s[1] & withiGambia2$site2==s[2]),]$LineWidth
        if(length(r) == 0)
            r = withiGambia2[which(withiGambia2$site1==s[2] & withiGambia2$site2==s[1]),]$LineWidth
        else
            lines(x=t$long[2000:40002], y=t$lat[2000:40002], col='steelblue', lwd=r)
    }
}

#for Brikama
brikamaAllFiles = list.files(path = paste0(relatednessDir,"/SiteConnectors"), pattern = "Brikama_", full.names = TRUE)
for(i in 1:length(brikamaAllFiles))
{
    if(i==1 | i==2 | i==5 | i==6 | i==7 | i==8)
    {
        s = gsub('.csv','',basename(brikamaAllFiles[i]))
        s = unlist(strsplit(s,'_'))
        t = fread(brikamaAllFiles[i])
        r = withiGambia2[which(withiGambia2$site1==s[1] & withiGambia2$site2==s[2]),]$LineWidth
        if(length(r) == 0)
            r = withiGambia2[which(withiGambia2$site1==s[2] & withiGambia2$site2==s[1]),]$LineWidth
        if( i==7)
            lines(x=t$long[2000:4002], y=t$lat[2000:4002], col='red', lwd=r)
        if(i==5 | i==6)
            lines(x=t$long[1:2000], y=t$lat[1:2000], col='steelblue', lwd=r)
        if(i==1)
            lines(x=t$long[2000:4002], y=t$lat[2000:4002], col='steelblue', lwd=r)
        else
            lines(x=t$long[2000:4002], y=t$lat[2000:4002], col='steelblue', lwd=r)
    }
    
}

#for dongoroBa
dongoroBaAllFiles = list.files(path = paste0(relatednessDir,"/SiteConnectors"), pattern = "DongoroBa_", full.names = TRUE)
for(i in 1:length(dongoroBaAllFiles))
{
    if(i==8)  #| i==10 | i==1 | i==11 | i==7 i==8 |
    {
        s = gsub('.csv','',basename(dongoroBaAllFiles[i]))
        s = unlist(strsplit(s,'_'))
        t = fread(dongoroBaAllFiles[i])
        r = withiGambia2[which(withiGambia2$site1==s[1] & withiGambia2$site2==s[2]),]$LineWidth
        if(length(r) == 0)
            r = withiGambia2[which(withiGambia2$site1==s[2] & withiGambia2$site2==s[1]),]$LineWidth
        lines(x=t$long[1:2000], y=t$lat[1:2000], col='steelblue', lwd=r)
    }
}

#for Njayel
njayelAllFiles = list.files(path = paste0(relatednessDir,"/SiteConnectors"), pattern = "Njayel_", full.names = TRUE)
for(i in 1:length(njayelAllFiles))
{
    if(i==10)  
    {
        s = gsub('.csv','',basename(njayelAllFiles[i]))
        s = unlist(strsplit(s,'_'))
        t = fread(njayelAllFiles[i])
        r = withiGambia2[which(withiGambia2$site1==s[1] & withiGambia2$site2==s[2]),]$LineWidth
        if(length(r) == 0)
            r = withiGambia2[which(withiGambia2$site1==s[2] & withiGambia2$site2==s[1]),]$LineWidth
        # if(i==6)
            lines(x=t$long[2000:4002], y=t$lat[2000:4002], col='steelblue', lwd=r)
        # else if(i==1)
        #     lines(x=t$long[1:2000], y=t$lat[1:2000], col='magenta', lwd=r)
        # else
        #     lines(x=t$long[2000:4002], y=t$lat[2000:4002], col='cyan', lwd=r)
    }
}
```

  
### **11. Relatedness in western Gambia **  
```{r echo=FALSE}
suppressPackageStartupMessages(expr = "")
require(data.table)
require(chorddiag)  #devtools::install_github("mattflor/chorddiag")
require(migest)
require(arcdiagram)  #devtools::install_github('gastonstat/arcdiagram')
require(edgebundleR)
require(igraph)
relatednessDir="/Users/km28/Documents/Karim/Karim/Gitlab/Transmission/RelatednessFolder"
rMatrix = readRDS(paste0(relatednessDir,'/rMatrixFrom_Raw.RDS'))
meta=fread(paste0(relatednessDir,"/SampleMetadata.txt"))
modifier = function(x){gsub('-','.',x)}
meta$SampleID=lapply(meta$SampleID,modifier)
rMatrix[is.na(rMatrix)]=0

easterGambia1 = raw[which(raw$site1=="Basse" | raw$site1=="Njayel" | raw$site1=="SareWuro"),]
easterGambia2 = easterGambia1[which(easterGambia1$site2=="Basse" | easterGambia1$site2=="Njayel" | easterGambia1$site2=="SareWuro"),]
rm(easterGambia1)
# easterGambia = easterGambia2[which(as.numeric(easterGambia2$`rMaf>30`)>0.4),]

gpsEast = readRDS(paste0(relatednessDir,'/EasternGambiaGPS.RDS'))
gpsEast = as.data.table(gpsEast)

connect = subset(easterGambia2, select=c(1,2,6))
names(connect) = c('from','to','value')
connect$value = as.numeric(connect$value)
# connect$value = 1
setkey(connect,"from")
setkey(gpsEast, "SampleID")
connect1 = connect[gpsEast, nomatch=0]
connect2 = subset(connect1, select=c(1,2,3,6,7,8,9))
names(connect2)[4:7]=c("villageCode1","compoundCode1","lat1","long1")
setkey(connect2,NULL); setkey(connect2,"to"); rm(connect, connect1)
connect3 = connect2[gpsEast, nomatch=0]
connect2 = subset(connect3, select=c(1:7,10:13))
names(connect2)[8:11]=c("villageCode2","compoundCode2","lat2","long2")
connect2 = connect2[which(connect2$value>0.2),]
connect2 = as.data.table(connect2)
setkeyv(connect2, c("villageCode1","villageCode2"))
npEast = connect2[,.N,by=c("villageCode1","villageCode2")]
kable(npEast, caption="#pairs in eastern Gambia with GPS data")
rm(connect3)
connect = subset(connect2, select=c(1:3))

#4. calculate the number of connection per sample
c( as.character(connect$from), as.character(connect$to)) %>%
    as_tibble() %>%
    group_by(value) %>%
    summarize(n=n()) -> coauth
colnames(coauth) <- c("name", "n")
coauth=as.data.table(coauth)
tmp=subset(connect2, select=c(1,4)); names(tmp)=c('iid','villageCode')
tmp1=subset(connect2, select=c(2,8)); names(tmp1)=c('iid','villageCode')
tmp = rbind(tmp,tmp1); rm(tmp1)
m = match(coauth$name, tmp$iid)
coauth$group=tmp$villageCode[m]

#8. Reorder dataset obtained after filtration and make the graph
coauth$grp = 0
coauth[which(coauth$group=='J'),]$grp=1
coauth[which(coauth$group=='K'),]$grp=2
coauth[which(coauth$group=='L'),]$grp=3
coauth[which(coauth$group=='M'),]$grp=4
coauth=subset(coauth, select = -3)
coauth = arrange(coauth, grp)
connect_difVill = connect #select links 

#10. re-ordering the connect data frame
for(i in 1:nrow(coauth)){
    if(i==1){
        connect_ForPlot = connect_difVill[which(connect_difVill$from==coauth$name[i]),]
    }else{
        if(nrow( connect_difVill[which(connect_difVill$from==coauth$name[i]),]) > 0){
            connect_ForPlot = rbind(connect_ForPlot, connect_difVill[which(connect_difVill$from==coauth$name[i]),])
        }
    }
}

connect_ForPlot$loc1=NA; connect_ForPlot$loc2=NA
m = match(connect_ForPlot$from, coauth$name)
connect_ForPlot$loc1 = coauth$grp[m]
m = match(connect_ForPlot$to, coauth$name)
connect_ForPlot$loc2 = coauth$grp[m]
connect_ForPlot$from = paste0(connect_ForPlot$loc1,'.',connect_ForPlot$from)
connect_ForPlot$to = paste0(connect_ForPlot$loc2,'.',connect_ForPlot$to)
connect_ForPlot$colour="black"
connect_ForPlot[which(connect_ForPlot$loc1 != connect_ForPlot$loc2),]$colour="green"
rel2 = subset(connect_ForPlot, select = c(1:2))

coauth1 = coauth
coauth1$name = paste0(coauth1$grp,'.',coauth1$name)

#11. drawing the graph
gg = graph.data.frame(rel2, directed=F, vertices=coauth1)
edge.start = ends(gg, es=E(gg), names = F)[,1]
edge.col = V(gg)$color[edge.start]

clr = as.factor(V(gg)$grp)
levels(clr) =  c("salmon", "orange", "lightskyblue", 'gray')
V(gg)$color = as.character(clr)
V(gg)$size = degree(gg)*2
E(gg)$width <- 7*connect_ForPlot$value
E(gg)$color = connect_ForPlot$colour
plot(gg, layout = layout.circle, vertex.label=NA)

edgebundle( gg, tension = 0.7 ) -> eb
eb


```









